File: domain/src/main/java/com/kaciula/archiman/domain/usecases/InitColdStart.java
package com.kaciula.archiman.domain.usecases;

import com.google.auto.value.AutoValue;
import com.kaciula.archiman.domain.boundary.data.AppInfoRepository;
import com.kaciula.archiman.domain.util.UseCase;
import io.reactivex.Completable;
import io.reactivex.Observable;

public class InitColdStart
    extends UseCase<InitColdStart.RequestModel, InitColdStart.ResponseModel> {

  private final AppInfoRepository appInfoRepository;

  public InitColdStart(AppInfoRepository appInfoRepository) {
    this.appInfoRepository = appInfoRepository;
  }

  @Override
  public Observable<ResponseModel> execute(final RequestModel requestModel) {
    return Completable.fromAction(() -> {
      int currentVersionCode = requestModel.currentVersionCode();
      if (appInfoRepository.isFirstTime()) {
        appInfoRepository.saveFirstTime(false);
        appInfoRepository.saveVersionCode(currentVersionCode);
      } else {
        if (appInfoRepository.getVersionCode() < currentVersionCode) {
          appInfoRepository.saveVersionCode(currentVersionCode);
        }
      }
    }).toObservable();
  }

  @AutoValue
  public abstract static class RequestModel implements UseCase.RequestModel {
    public abstract int currentVersionCode();

    public static RequestModel create(int currentVersionCode) {
      return new AutoValue_InitColdStart_RequestModel(currentVersionCode);
    }
  }


  @AutoValue
  abstract static class ResponseModel implements UseCase.ResponseModel {
    public static ResponseModel create() {
      return new AutoValue_InitColdStart_ResponseModel();
    }
  }
}

File: domain/src/main/kotlin/com/kaciula/archiman/domain/usecases/InitColdStart.kt
package com.kaciula.archiman.domain.usecases

import com.kaciula.archiman.domain.boundary.infrastructure.AppRepository
import com.kaciula.archiman.domain.usecases.util.CompletableUseCase
import com.kaciula.archiman.domain.util.Timberific
import io.reactivex.Completable

class InitColdStart(private val appRepository: AppRepository) : CompletableUseCase<InitColdStart.RequestModel>() {

    override fun execute(requestModel: RequestModel): Completable {
        return Completable.fromAction {
            Timberific.i("Initialize every cold start")
            val currentVersionCode = requestModel.currentVersionCode
            val app = appRepository.get()
            if (app.isFirstTime) {
                Timberific.i("First time running the app")
                app.isFirstTime = false
                app.versionCode = currentVersionCode
            } else {
                if (app.versionCode < currentVersionCode) {
                    Timberific.i("Old version code ${app.versionCode} is replaced with new version code $currentVersionCode")
                    app.versionCode = currentVersionCode
                } else {
                    Timberific.i("Just a basic cold start")
                }
            }
            appRepository.save(app)
        }
    }

    data class RequestModel(val currentVersionCode: Int)
}
